File: sdk/python/src/djtech_sdk/exceptions.py
"""
Exceptions for DJ Tech Agent Runtime SDK
"""

from typing import Any, Dict, Optional


class DJTechError(Exception):
    """Base exception for DJ Tech SDK"""

    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.message = message
        self.details = details or {}

    def __str__(self) -> str:
        if self.details:
            return f"{self.message} (details: {self.details})"
        return self.message


class APIError(DJTechError):
    """API returned an error response"""

    def __init__(
        self,
        message: str,
        status_code: int,
        code: Optional[str] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(message, details)
        self.status_code = status_code
        self.code = code


class AuthenticationError(APIError):
    """Authentication failed (401)"""

    def __init__(self, message: str = "Authentication failed", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 401, "AUTHENTICATION_ERROR", details)


class AuthorizationError(APIError):
    """Authorization failed (403)"""

    def __init__(self, message: str = "Access denied", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 403, "AUTHORIZATION_ERROR", details)


class NotFoundError(APIError):
    """Resource not found (404)"""

    def __init__(self, message: str = "Resource not found", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 404, "NOT_FOUND", details)


class ValidationError(APIError):
    """Request validation failed (400)"""

    def __init__(self, message: str = "Validation failed", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 400, "VALIDATION_ERROR", details)


class ConflictError(APIError):
    """Resource conflict (409)"""

    def __init__(self, message: str = "Resource conflict", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, 409, "CONFLICT", details)


class RateLimitError(APIError):
    """Rate limit exceeded (429)"""

    def __init__(
        self,
        message: str = "Rate limit exceeded",
        retry_after: Optional[int] = None,
        details: Optional[Dict[str, Any]] = None,
    ):
        super().__init__(message, 429, "RATE_LIMIT", details)
        self.retry_after = retry_after


class ServerError(APIError):
    """Server error (5xx)"""

    def __init__(self, message: str = "Internal server error", status_code: int = 500, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, status_code, "SERVER_ERROR", details)


class TimeoutError(DJTechError):
    """Request timeout"""

    def __init__(self, message: str = "Request timed out", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, details)


class NetworkError(DJTechError):
    """Network connection error"""

    def __init__(self, message: str = "Network error", details: Optional[Dict[str, Any]] = None):
        super().__init__(message, details)


def map_http_error(status_code: int, message: str, code: Optional[str] = None, details: Optional[Dict[str, Any]] = None) -> APIError:
    """Map HTTP status code to appropriate exception"""
    if status_code == 400:
        return ValidationError(message, details)
    elif status_code == 401:
        return AuthenticationError(message, details)
    elif status_code == 403:
        return AuthorizationError(message, details)
    elif status_code == 404:
        return NotFoundError(message, details)
    elif status_code == 409:
        return ConflictError(message, details)
    elif status_code == 429:
        return RateLimitError(message, details)
    elif 500 <= status_code < 600:
        return ServerError(message, status_code, details)
    else:
        return APIError(message, status_code, code, details)

File: sdk/typescript/src/exceptions.ts
/**
 * Exceptions for DJ Tech Agent Runtime SDK
 */

export class DJTechError extends Error {
  public readonly details: Record<string, unknown>;

  constructor(message: string, details: Record<string, unknown> = {}) {
    super(message);
    this.name = 'DJTechError';
    this.details = details;
    
    // Maintains proper stack trace in V8 environments
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, DJTechError);
    }
  }
}

export class APIError extends DJTechError {
  public readonly statusCode: number;
  public readonly code?: string;

  constructor(
    message: string,
    statusCode: number,
    code?: string,
    details: Record<string, unknown> = {}
  ) {
    super(message, details);
    this.name = 'APIError';
    this.statusCode = statusCode;
    this.code = code;
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, APIError);
    }
  }
}

export class AuthenticationError extends APIError {
  constructor(message = 'Authentication failed', details: Record<string, unknown> = {}) {
    super(message, 401, 'AUTHENTICATION_ERROR', details);
    this.name = 'AuthenticationError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AuthenticationError);
    }
  }
}

export class AuthorizationError extends APIError {
  constructor(message = 'Access denied', details: Record<string, unknown> = {}) {
    super(message, 403, 'AUTHORIZATION_ERROR', details);
    this.name = 'AuthorizationError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AuthorizationError);
    }
  }
}

export class NotFoundError extends APIError {
  constructor(message = 'Resource not found', details: Record<string, unknown> = {}) {
    super(message, 404, 'NOT_FOUND', details);
    this.name = 'NotFoundError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, NotFoundError);
    }
  }
}

export class ValidationError extends APIError {
  constructor(message = 'Validation failed', details: Record<string, unknown> = {}) {
    super(message, 400, 'VALIDATION_ERROR', details);
    this.name = 'ValidationError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ValidationError);
    }
  }
}

export class ConflictError extends APIError {
  constructor(message = 'Resource conflict', details: Record<string, unknown> = {}) {
    super(message, 409, 'CONFLICT', details);
    this.name = 'ConflictError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ConflictError);
    }
  }
}

export class RateLimitError extends APIError {
  public readonly retryAfter?: number;

  constructor(
    message = 'Rate limit exceeded',
    retryAfter?: number,
    details: Record<string, unknown> = {}
  ) {
    super(message, 429, 'RATE_LIMIT', details);
    this.name = 'RateLimitError';
    this.retryAfter = retryAfter;
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, RateLimitError);
    }
  }
}

export class ServerError extends APIError {
  constructor(message = 'Internal server error', statusCode = 500, details: Record<string, unknown> = {}) {
    super(message, statusCode, 'SERVER_ERROR', details);
    this.name = 'ServerError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ServerError);
    }
  }
}

export class TimeoutError extends DJTechError {
  constructor(message = 'Request timed out', details: Record<string, unknown> = {}) {
    super(message, details);
    this.name = 'TimeoutError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, TimeoutError);
    }
  }
}

export class NetworkError extends DJTechError {
  constructor(message = 'Network error', details: Record<string, unknown> = {}) {
    super(message, details);
    this.name = 'NetworkError';
    
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, NetworkError);
    }
  }
}

export function mapHttpError(
  statusCode: number,
  message: string,
  code?: string,
  details: Record<string, unknown> = {}
): APIError {
  switch (statusCode) {
    case 400:
      return new ValidationError(message, details);
    case 401:
      return new AuthenticationError(message, details);
    case 403:
      return new AuthorizationError(message, details);
    case 404:
      return new NotFoundError(message, details);
    case 409:
      return new ConflictError(message, details);
    case 429:
      return new RateLimitError(message, undefined, details);
    default:
      if (statusCode >= 500) {
        return new ServerError(message, statusCode, details);
      }
      return new APIError(message, statusCode, code, details);
  }
}